src/arkode/arkode_relaxation.c

```diff
@@ -0,0 +1,850 @@
+
+    /* Set defaults */
+    ark_mem->relax_mem->max_fails   = ARK_RELAX_DEFAULT_MAX_FAILS;
+    ark_mem->relax_mem->lower_bound = ARK_RELAX_DEFAULT_LOWER_BOUND;
+    ark_mem->relax_mem->upper_bound = ARK_RELAX_DEFAULT_UPPER_BOUND;
+    ark_mem->relax_mem->eta_fail    = ARK_RELAX_DEFAULT_ETA_FAIL;
+    ark_mem->relax_mem->solver      = ARK_RELAX_NEWTON;
+    ark_mem->relax_mem->res_tol     = ARK_RELAX_DEFAULT_RES_TOL;
+    ark_mem->relax_mem->rel_tol     = ARK_RELAX_DEFAULT_REL_TOL;
+    ark_mem->relax_mem->abs_tol     = ARK_RELAX_DEFAULT_ABS_TOL;
+    ark_mem->relax_mem->max_iters   = ARK_RELAX_DEFAULT_MAX_ITERS;
+
+    /* Initialize values */
+    ark_mem->relax_mem->relax_param_prev = ONE;
+
+    /* Update workspace sizes */
+    ark_mem->lrw += 12;
+    ark_mem->liw += 14;
+  }
+
+  /* Set function pointers */
+  ark_mem->relax_mem->relax_fn     = relax_fn;
+  ark_mem->relax_mem->relax_jac_fn = relax_jac_fn;
+  ark_mem->relax_mem->delta_y_fn   = delta_y_fn;
+  ark_mem->relax_mem->delta_e_fn   = delta_e_fn;
+  ark_mem->relax_mem->get_order_fn = get_order_fn;
+
+  /* Enable relaxation */
+  ark_mem->relax_enabled = SUNTRUE;
+
+  return ARK_SUCCESS;
+}
+
+/* Destructor called by driver */
+int arkRelaxDestroy(ARKodeRelaxMem relax_mem)
+{
+  if (!relax_mem) return ARK_SUCCESS;
+
+  /* Free structure */
+  free(relax_mem);
+
+  return ARK_SUCCESS;
+}
+
+/* Compute and apply relaxation, called by driver */
+int arkRelax(ARKodeMem ark_mem, int* relax_fails, realtype* dsm_inout,
+             int* nflag_out)
+{
+  int retval;
+  sunrealtype relax_val;
+  ARKodeRelaxMem relax_mem = ark_mem->relax_mem;
+
+  /* Get the relaxation memory structure */
+  if (!relax_mem)
+  {
+    arkProcessError(ark_mem, ARK_RELAX_MEM_NULL, "ARKODE",
+                    "arkRelax", MSG_RELAX_MEM_NULL);
+    return ARK_RELAX_MEM_NULL;
+  }
+
+  /* Compute the relaxation parameter */
+  retval = arkRelaxSolve(ark_mem, relax_mem, &relax_val);
+  if (retval < 0) return retval;
+  if (retval > 0)
+  {
+    /* Update failure counts */
+    relax_mem->num_fails++;
+    (*relax_fails)++;
+
+    /* Check for max fails in a step */
+    if (*relax_fails == relax_mem->max_fails) { return ARK_RELAX_FAIL; }
+
+    /* Return with an error if |h| == hmin */
+    if (SUNRabs(ark_mem->h) <= ark_mem->hmin * ONEPSM)
+    {
+      return ARK_RELAX_FAIL;
+    }
+
+    /* Return with error if using fixed step sizes */
+    if (ark_mem->fixedstep) { return(ARK_RELAX_FAIL); }
+
+    /* Cut step size and try again */
+    ark_mem->eta = relax_mem->eta_fail;
+
+#if SUNDIALS_LOGGING_LEVEL >= SUNDIALS_LOGGING_INFO
+    SUNLogger_QueueMsg(ARK_LOGGER, SUN_LOGLEVEL_INFO,
+                       "ARKODE::arkStep_TakeStep_Z", "relaxation",
+                       "relaxation failed");
+#endif
+
+    return TRY_AGAIN;
+  }
+
+  /* Update step size and error estimate */
+  ark_mem->h *= relax_val;
+  *dsm_inout *= SUNRpowerI(relax_val, relax_mem->get_order_fn(ark_mem));
```

Comment:
@drreynolds What do you think about adjusting `dsm` after relaxation? In particular when the step size might increase slightly.

Reply:
Given that there is a relatively tight upper bound on how much the step size can grow, I would think that it should be straightforward to estimate the resulting change in temporal error from the small increase.  If this potential error increase is only marginal then I think it should be fine as-is, particularly since it is statistically very rare that a step will pass with `dsm` identically equal to 1.  If the step could grow to a level where the resulting temporal error could change more significantly, then we could add something like a "bias" factor that is used in the temporal error test (e.g., a step would pass when ltebias*dsm <= 1); this "ltebias" factor could default to 1, but would be slightly larger when relaxation is used (where that factor falls out after performing the analysis above).